Mathlib/Algebra/Algebra/Operations.lean

```diff
@@ -0,0 +1,750 @@
+  rw [← span_pow, span_eq]
+#align submodule.pow_eq_span_pow_set Submodule.pow_eq_span_pow_set
+
+theorem pow_subset_pow {n : ℕ} : (↑M : Set A) ^ n ⊆ ↑(M ^ n : Submodule R A) :=
+  (pow_eq_span_pow_set M n).symm ▸ subset_span
+#align submodule.pow_subset_pow Submodule.pow_subset_pow
+
+theorem pow_mem_pow {x : A} (hx : x ∈ M) (n : ℕ) : x ^ n ∈ M ^ n :=
+  pow_subset_pow _ <| Set.pow_mem_pow hx _
+#align submodule.pow_mem_pow Submodule.pow_mem_pow
+
+theorem pow_toAddSubmonoid {n : ℕ} (h : n ≠ 0) : (M ^ n).toAddSubmonoid = M.toAddSubmonoid ^ n :=
+  by
+  induction' n with n ih
+  · exact (h rfl).elim
+  · rw [pow_succ, pow_succ, mul_toAddSubmonoid]
+    cases n with
+    | zero => rw [pow_zero, pow_zero, mul_one, ← mul_toAddSubmonoid, mul_one]
+    | succ n => rw [ih n.succ_ne_zero]
+#align submodule.pow_to_add_submonoid Submodule.pow_toAddSubmonoid
+
+theorem le_pow_toAddSubmonoid {n : ℕ} : M.toAddSubmonoid ^ n ≤ (M ^ n).toAddSubmonoid := by
+  obtain rfl | hn := Decidable.eq_or_ne n 0
+  · rw [pow_zero, pow_zero]
+    exact le_one_toAddSubmonoid
+  · exact (pow_toAddSubmonoid M hn).ge
+#align submodule.le_pow_to_add_submonoid Submodule.le_pow_toAddSubmonoid
+
+/-- Dependent version of `Submodule.pow_induction_on_left`. -/
+@[elab_as_elim]
+protected theorem pow_induction_on_left' {C : ∀ (n : ℕ) (x), x ∈ M ^ n → Prop}
+    (hr : ∀ r : R, C 0 (algebraMap _ _ r) (algebraMap_mem r))
+    (hadd : ∀ x y i hx hy, C i x hx → C i y hy → C i (x + y) (add_mem ‹_› ‹_›))
+    (hmul : ∀ m (hm : m ∈ M), ∀ (i x hx), C i x hx → C i.succ (m * x) (mul_mem_mul hm hx))
+    {x : A} {n : ℕ}
+    (hx : x ∈ M ^ n) : C n x hx := by
+  induction' n with n n_ih generalizing x
+  · rw [pow_zero] at hx
+    obtain ⟨r, rfl⟩ := hx
+    exact hr r
+  exact
+    Submodule.mul_induction_on' (fun m hm x ih => hmul _ hm _ _ _ (n_ih ih))
+      (fun x hx y hy Cx Cy => hadd _ _ _ _ _ Cx Cy) hx
+#align submodule.pow_induction_on_left' Submodule.pow_induction_on_left'
+
+/-- Dependent version of `Submodule.pow_induction_on_right`. -/
+@[elab_as_elim]
+protected theorem pow_induction_on_right' {C : ∀ (n : ℕ) (x), x ∈ M ^ n → Prop}
+    (hr : ∀ r : R, C 0 (algebraMap _ _ r) (algebraMap_mem r))
+    (hadd : ∀ x y i hx hy, C i x hx → C i y hy → C i (x + y) (add_mem ‹_› ‹_›))
+    (hmul :
+      ∀ i x hx, C i x hx →
+        ∀ m (hm : m ∈ M), C i.succ (x * m) ((pow_succ' M i).symm ▸ mul_mem_mul hx hm))
+    {x : A} {n : ℕ} (hx : x ∈ M ^ n) : C n x hx := by
+  induction' n with n n_ih generalizing x
+  · rw [pow_zero] at hx
+    obtain ⟨r, rfl⟩ := hx
+    exact hr r
+  revert hx
+  -- porting note: workaround for lean4#1926, was `simp_rw [pow_succ']`
+  have h_lean4_1926_aux : x ∈ M ^ Nat.succ n ↔ x ∈ M ^ n * M := by rw [pow_succ']
+  suffices h_lean4_1926 : ∀ (hx' : x ∈ M ^ n * M), C (Nat.succ n) x (h_lean4_1926_aux.mpr hx') from
+    fun hx => h_lean4_1926 (h_lean4_1926_aux.mp hx)
+  -- porting note: end workaround
```

Comment:
leanprover/lean4#1926. This workaround fails with a really strange message; is there something wrong with the hygiene of `suffices`?

Reply:
Fixed for now. Although it would be good to understand what went wrong.